refinery/file_store/models.py

```diff
@@ -169,17 +169,34 @@ def __unicode__(self):
 
     def save(self, *args, **kwargs):
         self.source = _map_source(self.source)
+
         # set file type using file extension
         try:
             extension = self.get_file_extension()
         except RuntimeError as exc:
-            logger.warn("Could not assign type to file '%s': %s", self, exc)
+            logger.warn("Could not assign type to file '%s': %s",
+                        self, exc)
         else:
             self.filetype = extension.filetype
-        # symlink datafile if necessary
-        if (not self.is_local() and os.path.isabs(self.source) and
-                settings.REFINERY_DATA_IMPORT_DIR not in self.source):
-            self._symlink_datafile()
+
+        if self.datafile:
+            # symlink datafile if necessary
+            if (not self.is_local() and os.path.isabs(self.source) and
+                    settings.REFINERY_DATA_IMPORT_DIR not in self.source):
+                self._symlink_datafile()
+        else:
+            # delete file
+            try:
+                old_instance = FileStoreItem.objects.get(pk=self.pk)
+            except FileStoreItem.DoesNotExist:
+                pass  # this is a newly created instance
+            except FileStoreItem.MultipleObjectsReturned as exc:
+                logger.critical(
+                    "Error retrieving FileStoreItem with primary key '%s': %s",
+                    self.pk, exc
+                )
+            else:
+                old_instance.delete_datafile(save_instance=False)
```

Comment:
This is tripping me up. Why are we worried about deleting a FileStoreItem's datafile here if we enter the else on 187 only if there isn't a datafile associated?